File: src/test/java/leetcode/LC053MaximumSubarray.java
package leetcode;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/maximum-subarray/
 */
public class LC053MaximumSubarray {
    public class Solution {
        public int maxSubArray(int[] nums) {
            int[] sums = new int[nums.length];
            int[] mins = new int[nums.length];
            mins[0] = 0;
            sums[0] = nums[0];
            for (int i = 1; i < nums.length; i++) {
                sums[i] = sums[i - 1] + nums[i];
                mins[i] = sums[i - 1] < mins[i - 1] ? sums[i - 1] : mins[i - 1];
            }
            int max = Integer.MIN_VALUE;
            for (int i = 0; i < nums.length; i++) {
                int delta = sums[i] - mins[i];
                max = delta > max ? delta : max;
            }
            return max;
        }
    }

    @Test
    public void test_1() throws Exception {
        assertEquals(1, new Solution().maxSubArray(new int[]{1}));
    }

    @Test
    public void test_Minus2_1() throws Exception {
        assertEquals(1, new Solution().maxSubArray(new int[]{-2, 1}));
    }

    @Test
    public void test_Minus2_Minus1() throws Exception {
        assertEquals(-1, new Solution().maxSubArray(new int[]{-2, -1}));
    }
}

File: algorithms/code/leetcode/lc053_maximum_subarray/lc053_maximum_subarray.py
# https://leetcode.com/problems/maximum-subarray/

import unittest


class Solution:
    def maxSubArray(self, nums: list[int]) -> int:
        summed = 0
        minimum = 0
        maximum = float('-inf')
        i = 0
        while i < len(nums):
            minimum = summed if summed < minimum else minimum
            summed += nums[i]
            delta = summed - minimum
            maximum = delta if delta > maximum else maximum
            i += 1
        return int(maximum)


class TestCode(unittest.TestCase):
    def test_example_1(self) -> None:
        assert Solution().maxSubArray([-2, 1, -3, 4, -1, 2, 1, -5, 4]) == 6

    def test_example_2(self) -> None:
        assert Solution().maxSubArray([1]) == 1

    def test_example_3(self) -> None:
        assert Solution().maxSubArray([5, 4, -1, 7, 8]) == 23

    def test_Minus2_1(self) -> None:
        assert Solution().maxSubArray([-2, 1]) == 1

    def test_Minus2_Minus1(self) -> None:
        assert Solution().maxSubArray([-2, -1]) == -1
